util/rwlock/src/rwlock.c

```diff
@@ -0,0 +1,163 @@
+#include "os/mynewt.h"
+#include "rwlock/rwlock.h"
+
+/**
+ * Unblocks the next pending user.  The caller must lock the mutex prior to
+ * calling this.
+ */
+static void
+rwlock_unblock(struct rwlock *lock)
+{
+    /* Give priority to pending writers. */
+    if (lock->pending_writers > 0) {
+        os_sem_release(&lock->wsem);
+        lock->pending_writers--;
+    } else {
+        while (lock->pending_readers > 0) {
+            os_sem_release(&lock->rsem);
+            lock->pending_readers--;
+        }
+    }
+}
+
+int
+rwlock_acquire_read(struct rwlock *lock, os_time_t timeout)
+{
+    bool acquired;
+    int rc;
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    /* Readers have low priority; if there is an active writer or a pending
+     * writer, we have to block.
+     */
+    if (lock->active_writer || lock->pending_writers > 0) {
+        lock->pending_readers++;
+        acquired = false;
+    } else {
+        lock->num_readers++;
+        acquired = true;
+    }
+
+    os_mutex_release(&lock->mtx);
+
+    if (acquired) {
+        /* No contention; lock acquired. */
+        return 0;
+    }
+
+    /* Wait for the lock to become available. */
+    rc = os_sem_pend(&lock->rsem, timeout);
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+    if (rc != 0 && rc != OS_NOT_STARTED) {
+        /* Semaphore timed out; undo increment and return error. */
+        lock->pending_readers--;
+        rc = SYS_ETIMEOUT;
+    } else {
+        /* Success; record reader ownership. */
+        lock->num_readers++;
+        rc = 0;
```

Comment:
This should go back to beginning to see if lock could be acquired. Writer might've grabbed the lock before reader task locked the mutex.